File: ddss/input.ts
import * as readline from "node:readline/promises";
import { stdin as input, stdout as output } from "node:process";
import { parse } from "atsds-bnf";
import { initializeDatabase, insertOrIgnore, Fact, Idea } from "./orm.js";
import { strRuleGetStrIdea, patchStdout } from "./utility.js";
import type { Sequelize } from "sequelize";

export async function main(addr: string, sequelize?: Sequelize) {
    if (!sequelize) {
        sequelize = await initializeDatabase(addr);
    }

    const rl = readline.createInterface({ input, output });
    rl.setPrompt("input: ");
    const unpatch = patchStdout(rl);

    try {
        rl.prompt(); // Initial prompt

        for await (const line of rl) {
            const data = line.trim();
            if (data === "" || data.startsWith("//")) {
                rl.prompt();
                continue;
            }

            try {
                const ds = parse(data);
                const dsStr = ds.toString();

                await insertOrIgnore(Fact, dsStr);
                const idea = strRuleGetStrIdea(dsStr);
                if (idea) {
                    await insertOrIgnore(Idea, idea);
                }
            } catch (e) {
                console.error(`error: ${e}`);
            }
            rl.prompt();
        }
    } catch (err) {
        // Silent catch for unexpected termination
    } finally {
        unpatch();
        rl.close();
    }
}

File: ddss/input.py
import asyncio
from prompt_toolkit import PromptSession
from prompt_toolkit.patch_stdout import patch_stdout
from apyds_bnf import parse
from .orm import insert_or_ignore, Facts, Ideas
from .utility import str_rule_get_str_idea


async def main(session):
    try:
        prompt = PromptSession()
        while True:
            try:
                with patch_stdout():
                    line = await prompt.prompt_async("input: ")
            except (EOFError, KeyboardInterrupt):
                raise asyncio.CancelledError()

            data = line.strip()
            if data == "":
                continue
            if data.startswith("//"):
                continue

            try:
                ds = parse(data)
            except Exception as e:
                print(f"error: {e}")
                continue

            async with session() as sess:
                await insert_or_ignore(sess, Facts, ds)
                if idea := str_rule_get_str_idea(ds):
                    await insert_or_ignore(sess, Ideas, idea)
                await sess.commit()
    except asyncio.CancelledError:
        pass
